packages/yew/src/platform/pinned/mpsc.rs

```diff
@@ -0,0 +1,347 @@
+
+        if let Some(m) = self.rx_waker.take() {
+            m.wake();
+        }
+    }
+}
+
+/// The receiver of an unbounded mpsc channel.
+#[derive(Debug)]
+pub struct UnboundedReceiver<T> {
+    inner: Rc<Inner<T>>,
+}
+
+impl<T> UnboundedReceiver<T> {
+    /// Try to read the next value from the channel.
+    ///
+    /// This function will return:
+    /// - `Ok(Some(T))` if a value is ready.
+    /// - `Ok(None)` if the channel has become closed.
+    /// - `Err(TryRecvError)` if the channel is not closed and the channel is empty.
+    pub fn try_next(&self) -> std::result::Result<Option<T>, TryRecvError> {
+        let inner = unsafe { &mut *self.inner.get_mut_unchecked() };
+
+        match (inner.items.pop_front(), inner.closed) {
+            (Some(m), _) => Ok(Some(m)),
+            (None, false) => Ok(None),
+            (None, true) => Err(TryRecvError {
+                _marker: PhantomData,
+            }),
+        }
+    }
+}
+
+impl<T> Stream for UnboundedReceiver<T> {
+    type Item = T;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let inner = unsafe { &mut *self.inner.get_mut_unchecked() };
+
+        match (inner.items.pop_front(), inner.closed) {
+            (Some(m), _) => Poll::Ready(Some(m)),
+            (None, false) => {
+                inner.rx_waker = Some(cx.waker().clone());
+                Poll::Pending
+            }
+            (None, true) => Poll::Ready(None),
+        }
+    }
+}
+
+impl<T> FusedStream for UnboundedReceiver<T> {
+    fn is_terminated(&self) -> bool {
+        let inner = unsafe { &mut *self.inner.get_mut_unchecked() };
+
+        inner.items.is_empty() && inner.closed
+    }
+}
+
+impl<T> Drop for UnboundedReceiver<T> {
+    fn drop(&mut self) {
+        let inner = unsafe { &mut *self.inner.get_mut_unchecked() };
+
+        inner.close();
+    }
+}
+
+/// The sender of an unbounded mpsc channel.
+#[derive(Debug)]
+pub struct UnboundedSender<T> {
+    inner: Rc<Inner<T>>,
+}
+
+impl<T> UnboundedSender<T> {
+    /// Sends a value to the unbounded receiver.
+    pub fn send_now(&self, item: T) -> Result<(), SendError<T>> {
+        let inner = unsafe { &mut *self.inner.get_mut_unchecked() };
+
+        if inner.closed {
+            return Err(SendError { inner: item });
+        }
+
+        inner.items.push_back(item);
+
+        if let Some(m) = inner.rx_waker.take() {
+            m.wake();
+        }
+
+        Ok(())
+    }
+
+    /// Closes the channel.
+    pub fn close_now(&self) {
+        let inner = unsafe { &mut *self.inner.get_mut_unchecked() };
+
+        inner.close();
+    }
+}
+
+impl<T> Clone for UnboundedSender<T> {
+    fn clone(&self) -> Self {
+        {
+            let inner = unsafe { &mut *self.inner.get_mut_unchecked() };
+            inner.sender_ctr += 1;
+        }
+
+        Self {
+            inner: self.inner.clone(),
+        }
```

Comment:
Almost certain this doesn't pass miri. Using a `&mut` as in line 162, 163 would invalidate all currently active const borrows, which is then used in `.clone()`. I think a lot of these could be fixed by stuffing `inner` into an `UnsafeCell`.

Reply:
This is due to that Miri does not consider that the this type is not `Send`.

We can continue to access this safely as we can guarantee that by the time of the mutable reference is acquired, no other code with immutable reference can run and hence wouldn't cause a racing condition.

Reply:
It's a stacked borrows issue. `&const` means that rust would be allowed to put reorder reads and omit redundant reads through the same reference, something that is violated when you also obtain a `&mut` to the same location. Hence, the reference can not be used after an mutable reference was unsafely crated. It has nothing to do with accessing this from another thread or `Send`.